docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+.. meta::
+  :description: This chapter describes the HIP programming model, the contract
+                between the programmer and the compiler/runtime executing the
+                code.
+  :keywords: AMD, ROCm, HIP, CUDA, C++ language extensions
+
+*******************************************************************************
+Programming Model
+*******************************************************************************
+
+HIP defines a model of mapping SIMT programs (Single Instruction, Multiple
+Threads) onto various architectures, primarily GPUs. While the model may be
+expressed in most imperative languages, (eg. Python via PyHIP) this document
+will focus on the original C/C++ API of HIP.
+
+Threading Model
+===============================================================================
+
+The SIMT nature of HIP is captured by the ability to execute user-provided
+device programs, expressed as single-source C/C++ functions or sources compiled
+online/offline to binaries in bulk.
+
+Multiple instances of the device program (aka. kernel) may execute in parallel,
+all uniquely identified by a set of integral values which are referred to as
+thread IDs. The set of integers identifying a thread relate to the hierarchy in
+which threads execute.
+
+.. _inherent_thread_model:
+
+Inherent Thread Model
+-------------------------------------------------------------------------------
+
+The thread hiearchy inherent to how AMD GPUs operate manifest as depicted in
+:numref:`inherent_thread_hierarchy`.
+
+.. _inherent_thread_hierarchy:
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside sets of uniform rectangles layered on oneanother
+        titled "Block". Each "Block" containing sets of uniform rectangles
+        layered on oneanother titled "Warp". Each of the "Warp" titled
+        rectangles filled with downward pointing arrows inside.
+
+  Hierarchy of thread groups.
+
+* The innermost grouping is called a warp, or a wavefront in ISA terms. A warp
+  is the most tightly coupled groups of threads, both physically and logically.
+
+  When referring to threads inside a warp, they may be called lanes, and the
+  integral value identifying them the lane ID. Lane IDs aren't quieried like
+  other thread IDs, but are user-calculated. As a consequence they are only as
+  multi-dimensional as the user interprets the calculated values to be.
+
+  The size of a warp is architecture dependent and always fixed. Warps are
+  signified by the set of communication primitives at their disposal, detailed
+  under :ref:`warp_cross_lane_functions`.
+
+* The middle grouping is called a block or thread block. The defining feature
+  of a block is that all threads in a block will share an instance of memory
+  which they may use to share data or synchronize with oneanother.
```

Comment:
In OpenCL it definitely is, and I'm not aware of it being any different in HIP. It may not be explicitly stated, but I'm not aware of runtimes being allowed at treating them otherwise: allowing a block to span multiple CUs (even when not using shared memory). The maximum block size is limited by the program counters of a given CU, which also happen to share the shared memory. (IMHO OpenCL does this better by explicitly nailing this down.)

> Compute Unit
>
>    An OpenCL device has one or more compute units. A work-group executes on a single compute unit. A compute unit is composed of one or more processing elements and local memory. A compute unit may also include dedicated texture filter units that can be accessed by its processing elements.
>
> Work-group
>
>    A collection of related work-items that execute on a single compute unit. The work-items in the group execute the same kernel-instance and share local memory and work-group functions.

Reply:
Changed the description - blocks need to run on a single CU. The need for shared memory is not good enough, then the whole grid could be seen as a block, as it shares the global memory, even if that would be slow. The possibility of synchronizing on a finer grained level is definitely more important, and that doesn't even necessarily require shared memory.